jupyterhub/app.py

```diff
@@ -2000,9 +2001,11 @@ async def init_roles(self):
                                 )
                         Class = orm.get_class(bearer)
                         orm_obj = Class.find(db, bname)
-                        roles.grant_role(
-                            db, entity=orm_obj, rolename=predef_role['name']
-                        )
+                        orm_role_bearers.append(orm_obj)
+                        # roles.grant_role(
+                        #     db, entity=orm_obj, rolename=predef_role['name']
+                        # )
+                setattr(predef_role_obj, bearer, orm_role_bearers)
         # make sure that on no admin situation, all roles are reset
         admin_role = orm.Role.find(db, name='admin')
         if not admin_role.users:
```

Comment:
I don't think this is the correct check for upgrading default roles. The check should be whether the admin role existed *at all* in `init_role_creation`. No admin users is itself a valid state, so it doesn't indicate an upgrade is taking place. Then we can set a private flag to consume here.

Reply:
Didn't consider that no admin users could be a valid situation but of course we can have other roles with elevated privileges as well, thanks